### ISSUE TITLE
large-scale perf testing

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#170]
In GitLab by @shanedsnyder on Nov 13, 2015, 16:04

We need to do some large-scale testing to compare the performance of Darshan 3.0 vs the current version of Darshan trunk (2.3.1). Cori would be a good system to test on, but could do Edison or Mira.

Some things we are interested in comparing:

- runtime overhead of wrapper functions
- runtime overhead of darshan init/shutdown procedures
- log file size for a couple different application workloads (in libz and bzip2 formats)
- maybe memory footprint

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Dec 8, 2015, 13:43

Valgrind has a nifty tool called Massif that can profile an application's heap memory allocations. Massif builds allocation trees to show which functions are responsible for the allocations and what percentage of total heap usage they cover. Specifically, we are concerned with how much memory Darshan functions are responsible for in a given application run. Massif ignores smaller allocations so this is mainly for looking at Darshan's larger data structures. 

For trunk version of Darshan:

- ~2.1 MiB allocated by Darshan at initialize time and persists throughout the application run
    - this mostly corresponds to the big runtime structure that stores file records and runtime records

For dev-modular version of Darshan:
- ~2.0 MiB allocated at initialize time, just like the trunk version
- ~2.5 MiB allocated by each the POSIX & MPI-IO modules when they are initialized
    - corresponds to module runtime structures and pre-allocated file record chunks (2 MiB per module)

So, the dev-modular branch does consume a good deal more memory than the master branch (about 5 MiB per process), but fortunately this is correctable:

* first, the 2 MiB the dev-modular branch allocates at initialize time is not necessary and was an oversight
    - specifically, this allocation is used for a staging buffer for Darshan's compression routines, which are only used at shutdown time; the allocation & lifetime of this buffer can be moved inside of Darshan's shutdown procedure so it doesn't hog memory from the application

* second, currently dev-modular version of darshan just lets each module allocate 2 MiB regardless of what they intend to store
    - we are working on redoing the memory allocation of darshan modules so they can request less than 2 MiB if they want, and also to enforce a hard maximum on the amount of memory all modules can use. With some tuning, we should be able to attain more comparable memory usage to trunk. 

I will keep this tool in mind as we continue development of the new memory allocation techniques, which we are investigating in the context of the mmap development branch. As we finalize that design, we can merge back into dev-modular and compare memory usage again to ensure Darshan isn't using much more than its configured max.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Dec 8, 2015, 13:47

In general, the dev-modular memory allocation strategy should still be fine for the time being as we don't have many modules that can hog a bunch of memory, but we will need to address at some point. I think ultimately the strategy of setting a hard per-process maximum memory usage and allowing modules to have more say in exactly how much memory they need will be sufficient.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Dec 14, 2015, 14:07

Would be good to turn this data into a tech report to share with others interested in Darshan overhead.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on May 3, 2016, 12:58

We have wrapped up a technical report evaluating Darshan 3.0.0 performance on Edison, a Cray XC30 supercomputer at NERSC. Our results indicate that the new Darshan version has little to no impact on the performance of instrumented applications. We also determined the time taken for Darshan to generate log files for an application and that the sizes of these logs are reasonable for full-time deployment on production HPC systems.

The technical report can be found [HERE](http://www.mcs.anl.gov/papers/MCS_TM_362_1.pdf).
